src/Makefile

```diff
@@ -130,52 +140,60 @@ depend: $(addsuffix /depend, $(SUBDIRS))
+$(SUBDIRS_LIB) : checkversion kaldi.mk mklibdir
+	$(MAKE) -C $@
 
-.PHONY: $(SUBDIRS)
-$(SUBDIRS) : checkversion kaldi.mk mklibdir
+.PHONY: bins $(SUBDIRS_BINS) libs
+$(SUBDIRS_BINS) : checkversion kaldi.mk mklibdir
 	$(MAKE) -C $@
 
 .PHONY: $(EXT_SUBDIRS)
 $(EXT_SUBDIRS) : checkversion kaldi.mk mklibdir ext_depend
 	$(MAKE) -C $@
 
-
 ### Dependency list ###
+# Three base libraries that mostly everything depends on
+BMU = base matrix util
+
 # this is necessary for correct parallel compilation
-#1)The tools depend on all the libraries
-bin fstbin gmmbin fgmmbin sgmm2bin featbin cudafeatbin nnetbin nnet2bin nnet3bin chainbin latbin ivectorbin lmbin kwsbin online2bin rnnlmbin cudadecoderbin: \
- base matrix util feat cudafeat tree gmm transform sgmm2 fstext hmm \
- lm decoder lat cudamatrix nnet nnet2 nnet3 ivector chain kws online2 rnnlm \
- cudadecoder
-
-#2)The libraries have inter-dependencies
-base: base/.depend.mk
+#1) The tools depend on all the libraries (which is a nonsense)
+bin chainbin featbin fgmmbin fstbin gmmbin ivectorbin \
+    kwsbin latbin lmbin nnet2bin nnet3bin nnetbin online2bin rnnlmbin sgmm2bin: \
+ $(BMU) chain cudamatrix decoder feat fstext gmm hmm ivector \
+        kws lat lm nnet nnet2 nnet3 online2 rnnlm sgmm2 transform tree
+
+#2) The libraries have inter-dependencies
+base:   base/.depend.mk
 matrix: base
-util: base matrix
-feat: base matrix util gmm transform tree
-tree: base util matrix
-gmm: base util matrix tree
-transform: base util matrix gmm tree
-sgmm2: base util matrix gmm tree transform hmm
-fstext: base util matrix tree
-hmm: base tree matrix util
-lm: base util matrix fstext
-decoder: base util matrix gmm hmm tree transform lat fstext
-lat: base util hmm tree matrix
-cudamatrix: base util matrix
-nnet: base util hmm tree matrix cudamatrix
-nnet2: base util matrix lat gmm hmm tree transform cudamatrix
-nnet3: base util matrix decoder lat gmm hmm tree transform cudamatrix chain fstext
-rnnlm: base util matrix cudamatrix nnet3 lm hmm
-chain: lat hmm tree fstext matrix cudamatrix util base
-ivector: base util matrix transform tree gmm
-#3)Dependencies for optional parts of Kaldi
-onlinebin: base matrix util feat tree gmm transform sgmm2 fstext hmm lm decoder lat cudamatrix nnet nnet2 online
-# python-kaldi-decoding: base matrix util feat tree gmm transform sgmm2 fstext hmm decoder lat online
-cudafeat: base matrix util gmm transform tree feat cudamatrix online2
-cudafeatbin: base matrix util gmm transform tree feat cudamatrix cudafeat online2
-online: decoder gmm transform feat matrix util base lat hmm tree
-online2: decoder gmm transform feat matrix util base lat hmm tree ivector cudamatrix nnet2 nnet3 chain
-kws: base util hmm tree matrix lat
-cudadecoder:  cudamatrix cudafeat online2 nnet3 ivector feat fstext lat chain transform
-cudadecoderbin: cudadecoder cudafeat cudamatrix online2 nnet3 ivector feat fstext lat chain transform
+util:   base matrix
+feat:      $(BMU) gmm transform tree
```

Comment:
This change might affect misc/maintenance/find_missing_dependencies.sh and reorder_addlibs.sh, not many people know about those scripts so it's not a big deal but just FYI.

Reply:
Ugh. As soon as I get a breather, I'll rewrite the build to use CMake. What's interesting, CMake itself warns you if you give it too many depedendencies that left unused. It also has a control to define how many times give a linker static libraries, so reordering is also not an issue (-llib1 -llib2 -llib1 -llib2 etc).

Fixing things like this one is a real PITA, and the more complex it gets, the more maintenance pain it is. The current effort started with a rewrite of everything top-down. I'd rather start bottom-up: Kaldi main codebase first, and then all the bells and whistles like automatically downloading dependencies etc.

CMake 2 used to be ugly. CMake post 3.14 ("dubbed "the modern CMake") is very different. And I compile CUDA code with CMake 3.19 natively (still need to call a package to find CUDA libraries, but that's true for C++ too), but NVCC is invoked with all defined flags, uses the defined C++ compiler etc.) And ninja is a much better build system for CMake: instead of a ton of unreadable Makefiles, it generates one huge unreadable file for ninja :) Surprisingly, ninja works even on Windows, too.

And CMake can be installed anywhere by untarring a tar, so we can just unpack any version we rely on in tools/

I tried many build systems, and the main take-away is that build is hard. Not that I like CMake, it has its own twists too, but currently it's probably the least painful and ugly build system.